cognite/client/data_classes/_base.py

```diff
@@ -100,6 +101,7 @@ def dump(self, camel_case: bool = False) -> Dict[str, Any]:
         Returns:
             Dict[str, Any]: A dictionary representation of the instance.
         """
+        assert not hasattr(self, "cognite_client")
```

Comment:
I'm not a big fan of asserts (also hasattr is relatively expensive) in non test code.

Maybe add a failing test instead?

Reply:
How will that pick up when a new subclass is added to the SDK? I guess we could potentially make a recursive `cls.__subclasses()` test... 😬 🤔 Yeah, I think I like that approach 👍 

Btw, I agree that asserts are incorrectly used all over the SDK; this one, however, should be very much in line with their intended use-case (i.e. for the SDK developer, _not the user_ (`-O`)).

Reply:
On second thought... no real way to verify all variants of subclasses. Some have required arguments, some are dataclasses with `__post_init__` method that do not allow inspection before instantiation... so I guess I'm back to wondering: _How will that pick up when a new subclass is added to the SDK?_

Reply:
I'm one of the people who thinks classes where a mistake in programming :D 

There is also always the option to just enforce it via code review :)

Reply:
# 🙈

Reply:
I went down the rabbit hole - figured the base Cognite classes have a lot of improvement potential. I removed the `hasattr` check from this PR, and will add tests (solved the issue above) in a separate PR!